kong/status/ready.lua

```diff
@@ -0,0 +1,111 @@
+local declarative = require "kong.db.declarative"
+local constants = require "kong.constants"
+
+local tonumber = tonumber
+local kong = kong
+
+local get_current_hash = declarative.get_current_hash
+
+
+local worker_count = ngx.worker.count()
+local kong_shm     = ngx.shared.kong
+
+local ngx_log        = ngx.log
+local ngx_WARN       = ngx.WARN
+
+local is_dbless = kong.configuration.database == "off"
+local is_control_plane = kong.configuration.role == "control_plane"
+
+local DECLARATIVE_PLUGINS_REBUILD_COUNT_KEY = 
+                                constants.DECLARATIVE_PLUGINS_REBUILD_COUNT_KEY
+local DECLARATIVE_ROUTERS_REBUILD_COUNT_KEY =
+                                constants.DECLARATIVE_ROUTERS_REBUILD_COUNT_KEY
+local DECLARATIVE_EMPTY_CONFIG_HASH = constants.DECLARATIVE_EMPTY_CONFIG_HASH
+
+
+local function is_dbless_ready(router_rebuilds, plugins_iterator_rebuilds)
+  if router_rebuilds < worker_count then
+    return false, "router rebuilds are not complete"
+  end
+
+  if plugins_iterator_rebuilds < worker_count then
+    return false, "plugins iterator rebuilds are not complete"
+  end
+
+  local current_hash = get_current_hash()
+
+  if not current_hash then
+    return false, "no configuration hash"
```

Comment:
```suggestion
    return false, "no configuration available"
```

The lack of hash indicates that configuration hasn't been synced to this Kong node.
This error message communicates to the end user that the Kong node doesn't have any configuration, which is a bit more helpful than 'no configuration hash' as we do not expect the user to understand "a missing hash means no configuration". The distinction is small but relevant because the hash is an implementation detail that can change in the future while the availability of configuration is unlikely to.

Reply:
You're right